src/wasm/wasm-type.cpp

```diff
@@ -2203,6 +2313,239 @@ bool FiniteShapeEquator::eq(const Rtt& a, const Rtt& b) {
   return a.depth == b.depth && eq(a.heapType, b.heapType);
 }
 
+size_t RecGroupHasher::operator()() const {
+  size_t digest = wasm::hash(group.size());
+  for (auto type : group) {
+    hash_combine(digest, topLevelHash(type));
+  }
+  return digest;
+}
+
+size_t RecGroupHasher::topLevelHash(HeapType type) const {
+  size_t digest = wasm::hash(type.isBasic());
+  if (type.isBasic()) {
+    wasm::rehash(digest, type.getID());
+  } else {
+    hash_combine(digest, hash(*getHeapTypeInfo(type)));
+  }
+  return digest;
+}
+
+size_t RecGroupHasher::innerHash(HeapType type) const {
+  // Do not recurse into the structure of this child type, but rather hash it as
+  // an index into a rec group. Only take the rec group identity into account if
+  // the child is not a member of the top-level group because in that case the
+  // group may not be canonicalized yet.
+  size_t digest = wasm::hash(type.getRecGroupIndex());
+  auto currGroup = type.getRecGroup();
+  if (currGroup != group) {
+    wasm::rehash(digest, currGroup.getID());
+  }
+  return digest;
+}
+
+size_t RecGroupHasher::hash(Type type) const {
+  size_t digest = wasm::hash(type.isBasic());
+  if (type.isBasic()) {
+    wasm::rehash(digest, type.getID());
+  } else {
+    hash_combine(digest, hash(*getTypeInfo(type)));
+  }
+  return digest;
+}
+
+size_t RecGroupHasher::hash(const TypeInfo& info) const {
+  size_t digest = wasm::hash(info.kind);
+  switch (info.kind) {
+    case TypeInfo::TupleKind:
+      hash_combine(digest, hash(info.tuple));
+      return digest;
+    case TypeInfo::RefKind:
+      rehash(digest, info.ref.nullable);
+      hash_combine(digest, innerHash(info.ref.heapType));
```

Comment:
Reading this, I am not clear on when `hash` is called vs `innerHash`, which I guess is the same question as above.

Reply:
Specifically, it is surprising to see hash on one and innerHash on another of the switch cases. I'd expect them all to be the same. I guess the issue is that HeapTypes get special treatment. But I wonder if `hash=>innerHash` might be less surprising when reading?

Reply:
Or, innerHash=hash could make sense too, really, with a comment.

Reply:
`hash` is for some intermediate structure (which is never a `HeapType`) and `innerHash` is for `HeapType`s. If this is causing confusion, I could also rename `innerHash` back to `hash`.

Reply:
Oh, I hadn't seen the last two comments when I wrote this. I'll rename `innerX` back to just `X`.